### ISSUE TITLE
Need help in implementation of cache buster in StealJS

### ISSUE THREAD

**Issue by soultrue** [stealjs/steal#1476]
HI,

I am using stealJS in my Canjs application and loading the files using below code code:

steal(
		'printonline/components/common',
		'printonline/common/properties.js',
		'printonline/common/constants.js',
		'printonline/common/resources.js',

I am using stealconfig.js to set the configuration for steal:
steal.config({
	map: {
		"*": {
			"can": "/templates/components/libraries/1.2/can/1.1.6/can",
		}
	},
	shim : {
		
		"/common/mustacheHelpers.js" : {
			deps: ["/templates/components/libraries/1.2/can/view/mustache/mustache.js"]
		}
	},
	ext: {
		js: "js",
		css: "css",
		mustache: "can/view/mustache"
	}
});

Now, as per our requirement, we have to load uncached version of all JS files so when the JS file load, it should load with appended string in path like abc.js?v=1234 something. I googled it and found one library called cache-buster(https://github.com/stealjs/cache-bust). I tried to implement it but I am not able to do that. Could you please guide me how can I do that?

I am using Steal 3.3. . I know in your latest version you have that feature by default but I can not update the StealJS version due to some limitation.

Please help!!

**Comment by matthewp** [comment]
Sounds like you are using legacy steal. cache-bust won't work with that. Honestly I'd probably just modify your `steal.js` file and inject a query string when it does the XHR request. Probably somewhere [around here](https://github.com/bitovi/legacy-steal/blob/9a826933d77ffa3fc22773d25d89819ce3ce5eee/steal.js#L167).

**Comment by soultrue** [comment]
Thanks Mattew!

I tried your solution and it works. Thanks!

But, it is only working for mustache files. JS files are still loading without query parameter.

steal.production.js
e.open("GET", a.src + "?v=1234", !(a.async === !1)) 

mustache:
registerInit.mustache?v=1234

JS:
registerModel.js (No query parameter)

Please suggest

**Comment by soultrue** [comment]
Thanks a lot Matthew!!

It works perfectly.

I can modify the stealjs in my local but in server, we do not have access of stealjs library. Only thing we can do is to modify steal.config.js. Is there a way to achieve the same thing by using the config file ?

Also, some one in our team suggesting that we can use webpack with canjs and do the same thing. what is your thought on that?

**Comment by matthewp** [comment]
You might be able to override `steal.idToUri` and append your querystring that way. https://github.com/bitovi/legacy-steal/blob/9a826933d77ffa3fc22773d25d89819ce3ce5eee/steal.js#L2427. You could probably do that inside of steal.config and it would work.

**Comment by soultrue** [comment]
Thanks!

I tried to implement your suggested approach and it is working however, I am getting  error on below section(**highlighted in Bold**):

steal.config({
	map: {
		"*": {
			"can": "/templates/components/libraries/1.2/can/1.1.6/can",
		}
	}
});

steal.idToUri = function( id, noJoin ) {
            // this is normalize
            var paths = **config.attr().paths** || {},
                path;
            // always run past
            **h.each(paths, function( part, replaceWith ) {**
                path = ""+id;
                // if path ends in / only check first part of id
                if((h.endsInSlashRegex.test(part) && path.indexOf(part) == 0) ||
                        // or check if its a full match only
                    path === part){
                    id = URI(path.replace(part, replaceWith));
                }
            })

            return noJoin ? id : config.attr().root.join(id)
}

Could you please suggest me on that. Thank you

**Comment by matthewp** [comment]
You put that in your steal.config? What I was thinking is you can wrap the call like:

```js
var idToUri = steal.idToUri;
steal.idToUri = function() {
  var url = idToUri.apply(this, arguments);

  // Now you can add your cache buster.
  return url + "?id=" + Math.random();
};
```

**Comment by soultrue** [comment]
Yes. I do not have access of to chnage stealjs as it is placed in other server. I can only access steal.config.js. That is why I overridden the steal function here.

**Comment by matthewp** [comment]
I understand. I'm suggesting you put the code I pasted into your steal.config.js. The code that you pasted won't work.

**Comment by soultrue** [comment]
Thanks a lot. You are great!!!

**Comment by matthewp** [comment]
Glad I could be of help!  Would you mind if I emailed you? I’d love to hear about your application and how you’re using steal/can.

**Comment by soultrue** [comment]
Sure. Thanks for your help again!

**Comment by matthewp** [comment]
What is a good email to contact you?

**Comment by soultrue** [comment]
Sorry for delay:

soultrue97@gmail.com
